src/tests/test_tls_session_manager.cpp

```diff
@@ -0,0 +1,1045 @@
+         }
+      return results;
+      };
+
+   return Test::flatten_result_lists(
+      {
+      CHECK_all("sessions expire", [&](auto, auto factory, auto& result)
+         {
+         auto mgr = factory();
+
+         auto handle = mgr->establish(default_session(Botan::TLS::Connection_Side::Server, cbs));
+         result.require("saved successfully", handle.has_value());
+         result.require("session was found", mgr->retrieve(handle.value(), cbs, plcy).has_value());
+         cbs.tick();
+         result.confirm("session has expired", !mgr->retrieve(handle.value(), cbs, plcy).has_value());
+         result.test_is_eq("session was deleted when it expired", mgr->remove_all(), size_t(0));
+         }),
+
+      CHECK_all("expired sessions are not found", [&](std::string type, auto factory, auto& result)
+         {
+         if(type == "Stateless")
+            return; // this manager can neither store nor find anything
+
+         auto mgr = factory();
+
+         auto handle_old = random_id();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs), handle_old);
+         result.require("session was found", mgr->retrieve(handle_old, cbs, plcy).has_value());
+
+         cbs.tick();
+         auto handle_new = random_id();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs), handle_new);
+         result.require("session was found", mgr->retrieve(handle_new, cbs, plcy).has_value());
+
+         auto sessions_and_handles = mgr->find(server_info, cbs, plcy);
+         result.require("sessions are found", !sessions_and_handles.empty());
+         result.test_is_eq("exactly one session is found", sessions_and_handles.size(), size_t(1));
+         result.test_is_eq("the new session is found", sessions_and_handles.front().second.id().value(), handle_new);
+
+         result.test_is_eq("old session was deleted when it expired", mgr->remove_all(), size_t(1));
+         }),
+
+      CHECK_all("session tickets are not reused", [&](std::string type, auto factory, auto& result)
+         {
+         if(type == "Stateless")
+            return; // this manager can neither store nor find anything
+
+         auto mgr = factory();
+
+         auto handle_1 = random_id();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs, Botan::TLS::Version_Code::TLS_V12), handle_1);
+         auto handle_2 = random_ticket();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs, Botan::TLS::Version_Code::TLS_V12), handle_2);
+
+#if defined(BOTAN_HAS_TLS_13)
+         auto handle_3 = random_id();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs, Botan::TLS::Version_Code::TLS_V13), handle_3);
+         auto handle_4 = random_ticket();
+         mgr->store(default_session(Botan::TLS::Connection_Side::Client, cbs, Botan::TLS::Version_Code::TLS_V13), handle_4);
+#endif
+
+         plcy.allow_session_reuse = false;
```

Comment:
Do we also somewhere test that the reuse of sessions works if the policy allows it?

Reply:
Not explicitly. Though, some tests above rely on this policy being `true` by default.